data_structures/rtree.cpp

```diff
@@ -67,6 +67,44 @@ void spatial::Rtree<T>::insert(Datum<T> const& new_datum) {
     }
 }
 
+/**
+ * Greedy k-NN query using distance browsing.
+ */
+template<typename T>
+std::vector<T> spatial::Rtree<T>::query_knn(
+    unsigned const k, coord_t const x, coord_t const y
+) const {
```

Comment:
```suggestion
template<typename Tree, typename Element>
std::vector<Element> query_knn(
    Tree<Element> const& tree, unsigned const k, coord_t const x, coord_t const y
) {
```
Very good, @gprender ! Indeed, it seems that the extra time/work spent on the quad tree paid off here in terms of producing an even better implementation of distance browsing quite quickly.

This suggestion isn't something to adopt on this report, but it is meant to nudge you to reflect on algorithmic abstractions now that you have the experience of implementing this for both the quad tree and the r-tree. Is this algorithm a property of each spatial index class, or is it an abstract routine that can be applied to any data structure that supports a particular interface? If the latter, what are the limitations of that abstraction?

Broadly in database query processing, when you think not so much of specific algorithms and particular trees, but of patterns that can be remixed, you have a way broader toolkit and better, quicker intuition of what approaches are best suited to particular contexts.